.github/workflows/delegator_low-perms_slsa3.yml

```diff
@@ -0,0 +1,279 @@
+      # during 'compilation': e.g., npm's install scripts.
+      contents: read # For repo checkout.
+    steps:
+      - name: debug
+        env:
+          TOKEN: ${{ toJson(needs.verify-token.outputs.slsa-verified-token) }}
+          RUNNER: ${{ fromJson(needs.verify-token.outputs.slsa-verified-token).builder.runner_label }}
+        run: |
+          echo "$TOKEN: $TOKEN"
+          echo "$RUNNER: $RUNNER"
+
+      - name: Checkout the tool repository
+        uses: slsa-framework/slsa-github-generator/.github/actions/secure-builder-checkout@main
+        with:
+          repository: ${{ needs.verify-token.outputs.tool-repository }}
+          ref: ${{ needs.verify-token.outputs.tool-ref }}
+          path: __TOOL_CHECKOUT_DIR__
+      - run: mv ./__TOOL_CHECKOUT_DIR__ ../__TOOL_CHECKOUT_DIR__
+
+      - name: Setup Action directory
+        env:
+          ACTION_PATH: ${{ fromJson(needs.verify-token.outputs.slsa-verified-token).tool.actions.build_artifacts.path }}
+        run: |
+          set -euo pipefail
+
+          mkdir -p ../__TOOL_ACTION_DIR__
+          mv ../__TOOL_CHECKOUT_DIR__/"$ACTION_PATH"/* ../__TOOL_ACTION_DIR__/
+
+          # Create the output directory.
+          mkdir "../$SLSA_OUTPUTS_DIR"
+
+          # Print for debugging
+          echo "ACTION_PATH=$ACTION_PATH"
+          tree
+
+      - name: Checkout the project repository
+        uses: slsa-framework/slsa-github-generator/.github/actions/secure-project-checkout@main
+
+      # NOTE: This calls the Action defined in the slsa-token.
+      - name: Build artifacts
+        id: build-artifacts-action
+        uses: ./../__TOOL_ACTION_DIR__
+        with:
+          slsa-workflow-inputs: ${{ toJson(fromJson(needs.verify-token.outputs.slsa-verified-token).tool.inputs) }}
+          slsa-layout-file: ../${{ env.SLSA_OUTPUTS_DIR }}/${{ env.SLSA_ARTIFACTS_FILE }}
+          slsa-workflow-secret1: ${{ secrets.secret1 }}
+          slsa-workflow-secret2: ${{ secrets.secret2 }}
+
+      - name: debug
+        env:
+          OUTPUTS: ${{ toJson(steps.build-artifacts-action.outputs) }}
+        run: |
+          echo "OUTPUTS: $OUTPUTS"
+
+      # NOTE: Needed to upload the file.
+      - name: Move artifact layout file to workspace
+        run: |
+          set -euo pipefail
+          mv "../${{ env.SLSA_OUTPUTS_DIR }}/${{ env.SLSA_ARTIFACTS_FILE }}" "${{ env.SLSA_ARTIFACTS_FILE }}"
+
+      - name: Upload artifact layout file
+        id: upload
+        uses: slsa-framework/slsa-github-generator/.github/actions/secure-upload-artifact@main
+        with:
+          name: "${{ needs.rng.outputs.value }}-${{ env.SLSA_ARTIFACTS_FILE }}"
+          path: "${{ env.SLSA_ARTIFACTS_FILE }}"
+
+  # generate-provenance generates and signs the provenance.
+  generate-provenance:
+    needs: [rng, verify-token, privacy-check, build-artifacts-ubuntu]
+    outputs:
+      attestations-download-name: "${{ needs.rng.outputs.value }}-slsa-attestations"
+    permissions:
```

Comment:
nit: add `contents: read` for private repositories

Reply:
Refresh my memory. Was `contents:read` needed for upload and download artifacts on private repos?

Reply:
ignore what I said. I remembered wrong.